src/app/WriteHandler.cpp

```diff
@@ -50,20 +48,10 @@ void WriteHandler::Shutdown()
 {
     VerifyOrReturn(mState != State::Uninitialized);
     mMessageWriter.Reset();
-    ClearExistingExchangeContext();
-    mpDelegate = nullptr;
+    mpExchangeCtx = nullptr;
```

Comment:
Shutdown is public.  Don't we still need logic here that conditionally closes `mpExchangeCtx`?  I don't understand how it is safe to assume another entity is responsible for this.

Reply:
@msandstedt `Shutdown` is public, but in practice should not be; the only real caller right now is internal to this class.  If we start handling writes asynchronously, that will change, and at that point we will need to be more careful about exactly when we close the exchange, and would also need to ensure the exchange stays open during the async bit...  I  think we have a few ways to proceed here:

1. Assume we will do write handling async at some point (even though there are no concrete plans for it right now that I am aware of), leave the currently-unused flexibility in place, and write code with that flexibility in mind.
2. Assume we might or might not do write handling async at some point, but in the meantime remove the various unused bits built around that premise, then readd them if we decide we do want to do async write handling.  This would mean making `Shutdown` private, for example, and updating the documentation accordingly.

I somewhat prefer option 2.  Within that, there are three sub-options: (a) do that cleanup in this PR, (b) do the cleanup in a followup, (c) do the cleanup in a PR that merges before this PR, then rebase on top.  I'd somewhat prefer (b) because this PR fixes bugs that are causing intermittent CI failures, but would be ok with (c) as well.

What would you prefer here?

Reply:
Since we are amidst cleanup and various fixes around the exchange manager, I would advocate for whatever solution is simpler and keeps public interfaces safe.  Option 2 seems like it fits with that.

As to whether we would do that cleanup now or later, I personally have a preference for now because that seems like it would more quickly lead to stability.  However, I don't want to stand in the way of merging something that can fix ongoing CI failures.

Reply:
I filed https://github.com/project-chip/connectedhomeip/issues/8651 and will fix it in the next day or two at most.